### PR TITLE
[develop] custom pipeline startup refactoring

### PR DESCRIPTION
**Pull request by star-e**
`pipeline.render()` is refactored into two parts: setup and render.
In setup step, we prepare a `cameraList` and setup the custom pipelines.
In render step,  the same `cameraList` is used in rendering.

Since pipeline setup is done in JS, in native pipeline, we will use event system to dispatch setup method to JS.
The `cameraList` is provided by Root through JSB.

### Changelog

* added `cameraList` to root.ts and root.h
* added `DIRECTOR_BEFORE_RENDER` event to setup pipeline in JS.
* pipeline now uses `interface` instead of `abstract class`. It is more light-weighted.

-------

### Continuous Integration

This pull request:

* [ ] needs automatic test cases check.
  > Manual trigger with `@cocos-robot run test cases` afterward.
* [ ] does not change any runtime related code or build configuration
  > If any reviewer thinks the CI checks are needed, please uncheck this option, then close and reopen the issue.

-------

### Compatibility Check

This pull request:

* [ ] changes public API, and have ensured backward compatibility with [deprecated features](https://github.com/cocos/cocos-engine/blob/v3.5.0/docs/contribution/deprecated-features.md).
* [ ] affects platform compatibility, e.g. system version, browser version, platform sdk version, platform toolchain, language version, hardware compatibility etc.
* [ ] affects file structure of the build package or build configuration which requires user project upgrade.
* [ ] **introduces breaking changes**, please list all changes, affected features and the scope of violation.

### CODE REVIEW

File: cocos/core/director.ts [hunk 1]
```diff
@@ -44,6 +44,8 @@ import { errorID, error, assertID, warnID } from './platform/debug';
 import { containerManager } from './memop/container-manager';
 import { uiRendererManager } from '../2d/framework/ui-renderer-manager';
 import { deviceManager } from '../gfx';
+import { PipelineBuilder } from '../rendering/custom/pipeline';
+import { macro } from './platform/macro';
```
Comment (star-e) [inline cocos/core/director.ts:48]: `interface PipelineBuilder` is used to build pipeline when custom pipeline is enabled.
Pipeline name is provided by `macro`.

File: cocos/core/director.ts [hunk 2]
```diff
@@ -166,6 +168,13 @@ export class Director extends EventTarget {
      */
     public static readonly EVENT_BEFORE_COMMIT = 'director_before_commit';
 
+    /**
+     * @en The event which will be triggered before the pipeline render.
+     * @zh 当前渲染帧渲染前所触发的事件。
+     * @event Director.EVENT_BEFORE_RENDER
+     */
+    public static readonly EVENT_BEFORE_RENDER = 'director_before_render';
+
```
Comment (star-e) [inline cocos/core/director.ts:177]: Add new event to call JS pipeline setup functioin.

File: cocos/core/director.ts [hunk 3]
```diff
@@ -731,6 +741,27 @@ export class Director extends EventTarget {
...
+    }
+
+    private setupRenderPipelineBuilder () {
+        if (this._root && this._root.usesCustomPipeline
+            && legacyCC.internal.createCustomPipeline
+            && legacyCC.internal.customPipelineBuilderMap) {
+            const map: Map<string, PipelineBuilder> = legacyCC.internal.customPipelineBuilderMap;
+            this._pipelineBuilder = map.get(macro.CUSTOM_PIPELINE_NAME) || null;
+            if (this._pipelineBuilder === null) {
+                this._pipelineBuilder = map.get('Forward')!;
```
Comment (star-e) [inline cocos/core/director.ts:759]: Find required custom pipeline builder when customization is enabled. If pipeline not found, use builtin 'Forward' pipeline.

File: cocos/core/root.jsb.ts [hunk 4]
```diff
@@ -183,6 +183,10 @@ rootProto._onDirectorBeforeCommit = function () {
     legacyCC.director.emit(legacyCC.Director.EVENT_BEFORE_COMMIT);
 };
 
+rootProto._onDirectorBeforeRender = function () {
+    legacyCC.director.emit(legacyCC.Director.EVENT_BEFORE_RENDER);
+};
```
Comment (star-e) [inline cocos/core/root.jsb.ts:188]: Dispatch JS callback from c++

File: cocos/core/root.ts [hunk 5]
```diff
@@ -234,6 +234,10 @@ export class Root {
         return this._useDeferredPipeline;
     }
 
+    public get cameraList (): Camera[] {
+        return this._cameraList;
+    }
```
Comment (star-e) [inline cocos/core/root.ts:239]: cameraList is kept and shared between pipeline setup and render.

File: cocos/core/root.ts [hunk 6]
```diff
@@ -484,7 +489,9 @@ export class Root {
         }
 
         const windows = this._windows;
-        const cameraList: Camera[] = [];
+        const cameraList = this._cameraList;
+        cameraList.splice(0, cameraList.length);
+
```
Comment (star-e) [inline cocos/core/root.ts]: cameraList is rebuilt every frame.

**Comment by star-e** [comment]
@cocos-robot run test cases

File: cocos/core/root.ts (hunk 6 above)
Comment (dumganhar) [reply cocos/core/root.ts]: Could we just use `cameraList.length = 0;` ?

**Review by dumganhar — APPROVED** [review]

File: cocos/core/root.ts (hunk 6 above)
Comment (star-e) [reply cocos/core/root.ts]: Yes. I forget we can do this.

**Review by minggo — APPROVED** [review]

**Pull request merged**